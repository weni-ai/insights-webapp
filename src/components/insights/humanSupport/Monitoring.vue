<template>
  <section class="monitoring">
    <h1>Monitoring</h1>
  </section>
</template>

<script setup lang="ts"></script>

<style scoped lang="scss">
.monitoring {
  display: flex;
  flex-direction: column;
  gap: $unnnic-space-8;
}
</style>
