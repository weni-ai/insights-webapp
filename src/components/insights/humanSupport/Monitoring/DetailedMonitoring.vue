<template>
  <section class="detailed-monitoring">
    <p class="detailed-monitoring__title">
      {{ $t('human_support_dashboard.detailed_monitoring.title') }}
    </p>
    <Transition name="filters-fade">
      <section
        v-if="['attendant', 'pauses'].includes(activeTabName)"
        class="detailed-monitoring__filters"
      >
        <DetailedFilters />
      </section>
    </Transition>
    <section class="detailed-monitoring__tabs">
      <UnnnicTab
        data-testid="human-support-tab"
        :tabs="tabsKeys"
        :activeTab="activeTabName"
        @change="changeActiveTabName"
      >
        <template
          v-for="[key, tab] in Object.entries(tabs)"
          #[`tab-head-${key}`]
          :key="`tab-head-${key}`"
        >
          {{
            $t(`human_support_dashboard.detailed_monitoring.tabs.${tab.name}`)
          }}
        </template>
        <template
          v-for="key in Object.keys(tabs)"
          #[`tab-panel-${key}`]
          :key="`tab-panel-${key}`"
        >
          <component
            :is="tabs[key].component"
            :data-testid="`tab-panel-${key}`"
          />
        </template>
      </UnnnicTab>
    </section>
  </section>
</template>

<script setup lang="ts">
import { UnnnicTab } from '@weni/unnnic-system';
import InAwaiting from './Tables/InAwaiting.vue';
import InProgress from './Tables/InProgress.vue';
import Attendant from './Tables/Attendant.vue';
import Pauses from './Tables/Pauses.vue';
import { ref } from 'vue';
import DetailedFilters from './Filters/DetailedFilters.vue';

const tabs = {
  in_awaiting: {
    name: 'in_awaiting',
    component: InAwaiting,
  },
  in_progress: {
    name: 'in_progress',
    component: InProgress,
  },
  attendant: {
    name: 'attendant',
    component: Attendant,
  },
  pauses: {
    name: 'pauses',
    component: Pauses,
  },
};

const tabsKeys = Object.keys(tabs);

const activeTabName = ref('in_progress');

const changeActiveTabName = (tab: string) => {
  activeTabName.value = tab;
};
</script>

<style scoped lang="scss">
.detailed-monitoring {
  display: flex;
  padding: $unnnic-space-6;
  flex-direction: column;
  gap: $unnnic-space-6;

  border-radius: $unnnic-radius-2;
  border: 1px solid $unnnic-color-neutral-soft;
  background: $unnnic-color-background-white;

  &__title {
    font: $unnnic-font-display-2;
  }

  &__filters {
    display: flex;
    flex-direction: column;
    flex: 0 0 calc(100% / 4);
    max-width: calc(100% / 4);
  }
}

.filters-fade-enter-active,
.filters-fade-leave-active {
  transition: all 0.3s ease;
}

.filters-fade-enter-from {
  opacity: 0;
  transform: translateY(-10px);
}

.filters-fade-leave-to {
  opacity: 0;
  transform: translateY(-10px);
}
</style>
