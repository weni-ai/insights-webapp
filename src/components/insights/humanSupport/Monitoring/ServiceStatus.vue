<template>
  <section
    class="service-status"
    data-testid="service-status"
  >
    <p
      class="service-status__title"
      data-testid="service-status-title"
    >
      {{ $t('human_support_dashboard.support_status.title') }}
    </p>
    <section
      class="service-status__cards"
      data-testid="service-status-cards"
    >
      <template
        v-for="(card, index) in cardDefinitions"
        :key="`${card.id}-${index}`"
      >
        <CardConversations
          class="service-status__card"
          :title="$t(card.titleKey)"
          :value="getCardValue(card.id)"
          :tooltipInfo="$t(card.tooltipKey)"
          borderRadius="full"
          :tooltipSide="getTooltipSide(index)"
          :isLoading="isLoadingCards"
        />
      </template>
    </section>
  </section>
</template>

<script setup lang="ts">
import { onMounted, computed } from 'vue';

import CardConversations from '@/components/insights/cards/CardConversations.vue';
import { useHumanSupportMonitoring } from '@/store/modules/humanSupport/monitoring';
import { storeToRefs } from 'pinia';

type CardId = 'is_awaiting' | 'in_progress' | 'finished';

interface CardData {
  id: CardId;
  titleKey: string;
  tooltipKey: string;
}

const cardDefinitions: CardData[] = [
  {
    id: 'is_awaiting',
    titleKey: 'human_support_dashboard.support_status.is_awaiting',
    tooltipKey: 'human_support_dashboard.support_status.tooltips.is_awaiting',
  },
  {
    id: 'in_progress',
    titleKey: 'human_support_dashboard.support_status.is_in_progress',
    tooltipKey:
      'human_support_dashboard.support_status.tooltips.is_in_progress',
  },
  {
    id: 'finished',
    titleKey: 'human_support_dashboard.support_status.is_closed',
    tooltipKey: 'human_support_dashboard.support_status.tooltips.is_closed',
  },
];

const humanSupportMonitoring = useHumanSupportMonitoring();
const { loadServiceStatusData } = humanSupportMonitoring;
const { serviceStatusData, loadingServiceStatusData } = storeToRefs(
  humanSupportMonitoring,
);

const isLoadingCards = computed(() => loadingServiceStatusData.value);

const getCardValue = (id: CardId) => {
  return serviceStatusData.value[id]?.toString() || '-';
};

const getTooltipSide = (index: number) => {
  const lastIndex = cardDefinitions.length - 1;

  if (lastIndex === index) return 'left';

  return 'top';
};

onMounted(() => {
  loadServiceStatusData();
});
</script>

<style scoped lang="scss">
$min-height: 112px;

.service-status {
  display: flex;
  flex-direction: column;
  gap: $unnnic-space-3;

  &__title {
    font: $unnnic-font-display-2;
  }

  &__cards {
    display: flex;
    gap: $unnnic-space-4;
    min-height: $min-height;
  }
}
</style>
