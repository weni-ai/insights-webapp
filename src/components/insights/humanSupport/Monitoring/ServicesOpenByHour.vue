<template>
  <section
    class="services-open-by-hour"
    data-testid="services-open-by-hour"
  >
    <LineChart
      data-testid="services-open-by-hour-chart"
      :title="$t('human_support_dashboard.services_open_by_hour.title')"
      :chartData="data"
      :seeMore="false"
      :isLoading="isLoading"
    />
  </section>
</template>

<script setup lang="ts">
import { onMounted, computed } from 'vue';
import { useHumanSupportMonitoring } from '@/store/modules/humanSupport/monitoring';
import LineChart from '@/components/insights/charts/LineChart.vue';
import { storeToRefs } from 'pinia';

const humanSupportMonitoring = useHumanSupportMonitoring();
const { loadHumanSupportByHourData } = humanSupportMonitoring;

const { servicesOpenByHourData, loadingHumanSupportByHourData } = storeToRefs(
  humanSupportMonitoring,
);

const data = computed(() => {
  const formattedData = {
    labels: servicesOpenByHourData.value.map((item) => item.label),
    datasets: [
      {
        data: servicesOpenByHourData.value.map((item) => item.value),
      },
    ],
  };

  return formattedData;
});

const isLoading = computed(() => {
  return loadingHumanSupportByHourData.value;
});

onMounted(() => {
  loadHumanSupportByHourData();
});
</script>

<style scoped lang="scss">
.services-open-by-hour {
  display: flex;
  flex-direction: column;
  gap: $unnnic-space-3;
  max-height: 250px;
}
</style>
