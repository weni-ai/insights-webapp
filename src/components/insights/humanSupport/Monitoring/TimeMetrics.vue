<template>
  <section class="time-metrics">
    <p class="time-metrics__title">
      {{ $t('human_support_dashboard.time_metrics.title') }}
    </p>
    <section class="time-metrics__cards">
      <template
        v-for="(card, index) in cardDefinitions"
        :key="`${card.id}-${index}`"
      >
        <CardConversations
          class="time-metrics__card"
          :title="$t(card.titleKey)"
          :value="getCardValue(card.id)"
          :description="getCardSubValue(card.id)"
          :tooltipInfo="$t(card.tooltipKey)"
          borderRadius="full"
          activeDescriptionGap
          enableTimeIcon
          :tooltipSide="getTooltipSide(index)"
          :isLoading="isLoadingCards"
        />
      </template>
    </section>
  </section>
</template>

<script setup lang="ts">
import { onMounted, computed } from 'vue';

import CardConversations from '@/components/insights/cards/CardConversations.vue';
import { useHumanSupportMonitoring } from '@/store/modules/humanSupport/monitoring';
import { storeToRefs } from 'pinia';
import { useI18n } from 'vue-i18n';
import { formatSecondsToTime } from '@/utils/time';

type CardId =
  | 'average_time_is_waiting'
  | 'average_time_first_response'
  | 'average_time_chat';

interface CardData {
  id: CardId;
  titleKey: string;
  tooltipKey: string;
}

const cardDefinitions: CardData[] = [
  {
    id: 'average_time_is_waiting',
    titleKey: 'human_support_dashboard.time_metrics.average_time_is_waiting',
    tooltipKey:
      'human_support_dashboard.time_metrics.tooltips.average_time_is_waiting',
  },
  {
    id: 'average_time_first_response',
    titleKey:
      'human_support_dashboard.time_metrics.average_time_first_response',
    tooltipKey:
      'human_support_dashboard.time_metrics.tooltips.average_time_first_response',
  },
  {
    id: 'average_time_chat',
    titleKey: 'human_support_dashboard.time_metrics.average_time_chat',
    tooltipKey:
      'human_support_dashboard.time_metrics.tooltips.average_time_chat',
  },
];

const { t } = useI18n();

const humanSupportMonitoring = useHumanSupportMonitoring();
const { loadTimeMetricsData } = humanSupportMonitoring;
const { timeMetricsData, loadingTimeMetricsData } = storeToRefs(
  humanSupportMonitoring,
);

const isLoadingCards = computed(() => loadingTimeMetricsData.value);

const getCardValue = (id: CardId) => {
  const data = timeMetricsData.value[id];
  return formatSecondsToTime(data?.average);
};

const getCardSubValue = (id: CardId) => {
  const data = timeMetricsData.value[id];
  if (data?.max !== null && data?.max !== undefined) {
    return `${t('human_support_dashboard.time_metrics.max')}: ${formatSecondsToTime(data.max)}`;
  }
  return '';
};

const getTooltipSide = (index: number) => {
  const lastIndex = cardDefinitions.length - 1;

  if (lastIndex === index) return 'left';

  return 'top';
};

onMounted(() => {
  loadTimeMetricsData();
});
</script>

<style scoped lang="scss">
$min-height: 134px;

.time-metrics {
  display: flex;
  flex-direction: column;
  gap: $unnnic-space-3;

  &__title {
    font: $unnnic-font-display-2;
  }

  &__cards {
    display: flex;
    gap: $unnnic-space-4;
    min-height: $min-height;
  }
}
</style>
