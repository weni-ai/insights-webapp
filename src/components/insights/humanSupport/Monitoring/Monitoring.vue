<template>
  <section class="monitoring">
    <ServiceStatus />
    <TimeMetrics />
  </section>
</template>

<script setup lang="ts">
import { onMounted, onUnmounted, ref } from 'vue';

import { useDashboards } from '@/store/modules/dashboards';
import ServiceStatus from './ServiceStatus.vue';
import TimeMetrics from './TimeMetrics.vue';

const isLoading = ref(false);
let autoRefreshInterval: ReturnType<typeof setInterval> | null = null;

const AUTO_REFRESH_INTERVAL = 60 * 1000;

const { updateLastUpdatedRequest } = useDashboards();

const loadData = async () => {
  if (isLoading.value) {
    return;
  }

  try {
    isLoading.value = true;
    updateLastUpdatedRequest();
  } catch (error) {
    console.error('Erro to get data:', error);
  } finally {
    isLoading.value = false;
  }
};

const startAutoRefresh = () => {
  if (autoRefreshInterval) {
    clearInterval(autoRefreshInterval);
  }

  autoRefreshInterval = setInterval(() => {
    loadData();
  }, AUTO_REFRESH_INTERVAL);
};

const stopAutoRefresh = () => {
  if (autoRefreshInterval) {
    clearInterval(autoRefreshInterval);
    autoRefreshInterval = null;
  }
};

onMounted(() => {
  console.log('Monitoring mounted');

  loadData();

  startAutoRefresh();
});

onUnmounted(() => {
  console.log('Monitoring unmounted');

  stopAutoRefresh();
});
</script>

<style scoped lang="scss">
.monitoring {
  display: flex;
  flex-direction: column;
  gap: $unnnic-space-8;
}
</style>
