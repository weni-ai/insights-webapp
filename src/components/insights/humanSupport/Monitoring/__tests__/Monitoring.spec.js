import { beforeEach, describe, expect, it, vi } from 'vitest';
import { config, mount } from '@vue/test-utils';
import { createI18n } from 'vue-i18n';

import Monitoring from '../Monitoring.vue';

const mockDashboardStore = {
  updateLastUpdatedRequest: vi.fn(),
};

vi.mock('@/store/modules/dashboards', () => ({
  useDashboards: () => mockDashboardStore,
}));

vi.mock('pinia', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    storeToRefs: (store) => store,
  };
});

const i18n = createI18n({
  legacy: false,
  locale: 'en',
  messages: {
    en: {},
  },
  fallbackWarn: false,
  missingWarn: false,
});

config.global.plugins = [i18n];

describe('Monitoring', () => {
  let wrapper;

  const createWrapper = (storeOverrides = {}) => {
    Object.assign(mockDashboardStore, storeOverrides);
    return mount(Monitoring, {
      global: {
        stubs: {
          ServiceStatus: true,
          TimeMetrics: true,
          ServicesOpenByHour: true,
        },
      },
    });
  };

  const section = () => wrapper.find('[data-testid="monitoring"]');

  beforeEach(() => {
    vi.clearAllMocks();
    vi.useFakeTimers();
    vi.spyOn(console, 'log').mockImplementation(() => {});
    vi.spyOn(console, 'error').mockImplementation(() => {});

    Object.assign(mockDashboardStore, {
      updateLastUpdatedRequest: vi.fn(),
    });

    wrapper = createWrapper();
  });

  afterEach(() => {
    vi.clearAllTimers();
    vi.restoreAllMocks();
  });

  describe('Component rendering', () => {
    it('should render main section', () => {
      expect(section().exists()).toBe(true);
    });

    it('should render both child components', () => {
      expect(wrapper.findComponent({ name: 'ServiceStatus' }).exists()).toBe(
        true,
      );
      expect(wrapper.findComponent({ name: 'TimeMetrics' }).exists()).toBe(
        true,
      );
      expect(
        wrapper.findComponent({ name: 'ServicesOpenByHour' }).exists(),
      ).toBe(true);
    });

    it('should have correct data-testids for child components', () => {
      const serviceStatus = wrapper.findComponent({ name: 'ServiceStatus' });
      const timeMetrics = wrapper.findComponent({ name: 'TimeMetrics' });
      const servicesOpenByHour = wrapper.findComponent({
        name: 'ServicesOpenByHour',
      });

      expect(serviceStatus.attributes('data-testid')).toBe(
        'monitoring-service-status',
      );
      expect(timeMetrics.attributes('data-testid')).toBe(
        'monitoring-time-metrics',
      );
      expect(servicesOpenByHour.attributes('data-testid')).toBe(
        'monitoring-services-open-by-hour',
      );
    });
  });

  describe('Component structure', () => {
    it('should have correct CSS classes', () => {
      expect(section().classes()).toContain('monitoring');
    });
  });

  describe('Lifecycle management', () => {
    it('should initialize with correct default values', () => {
      const vm = wrapper.vm;

      expect(vm.isLoading).toBe(false);
      expect(vm.autoRefreshInterval).not.toBe(null);
      expect(vm.AUTO_REFRESH_INTERVAL).toBe(60000);
    });

    it('should call loadData on mount', () => {
      expect(mockDashboardStore.updateLastUpdatedRequest).toHaveBeenCalled();
      expect(console.log).toHaveBeenCalledWith('Monitoring mounted');
    });

    it('should start auto refresh on mount', () => {
      const vm = wrapper.vm;
      expect(vm.autoRefreshInterval).not.toBe(null);
    });

    it('should cleanup on unmount', () => {
      const vm = wrapper.vm;
      wrapper.unmount();

      expect(console.log).toHaveBeenCalledWith('Monitoring unmounted');
      expect(vm.autoRefreshInterval).toBe(null);
    });
  });

  describe('Data loading', () => {
    it('should handle loadData correctly', async () => {
      const vm = wrapper.vm;

      expect(vm.isLoading).toBe(false);

      await vm.loadData();
      expect(mockDashboardStore.updateLastUpdatedRequest).toHaveBeenCalled();
    });

    it('should prevent concurrent loads', async () => {
      const vm = wrapper.vm;

      vm.isLoading = true;
      const loadResult = await vm.loadData();

      expect(loadResult).toBeUndefined();
    });

    it('should handle errors gracefully', async () => {
      const vm = wrapper.vm;
      const error = new Error('Load error');

      mockDashboardStore.updateLastUpdatedRequest.mockImplementationOnce(() => {
        throw error;
      });

      await vm.loadData();

      expect(console.error).toHaveBeenCalledWith('Erro to get data:', error);
      expect(vm.isLoading).toBe(false);
    });
  });

  describe('Auto refresh functionality', () => {
    it('should start auto refresh with correct interval', () => {
      const vm = wrapper.vm;

      vm.startAutoRefresh();

      expect(vm.autoRefreshInterval).not.toBe(null);
    });

    it('should execute loadData on interval', () => {
      const vm = wrapper.vm;

      const originalSetInterval = globalThis.setInterval;
      const mockCallback = vi.fn();
      globalThis.setInterval = vi.fn(() => {
        mockCallback();
        return 123;
      });

      vm.startAutoRefresh();

      expect(globalThis.setInterval).toHaveBeenCalledWith(
        expect.any(Function),
        60000,
      );

      globalThis.setInterval = originalSetInterval;
    });

    it('should stop auto refresh correctly', () => {
      const vm = wrapper.vm;

      vm.startAutoRefresh();
      expect(vm.autoRefreshInterval).not.toBe(null);

      vm.stopAutoRefresh();
      expect(vm.autoRefreshInterval).toBe(null);
    });

    it('should handle multiple stop calls gracefully', () => {
      const vm = wrapper.vm;

      expect(() => {
        vm.stopAutoRefresh();
        vm.stopAutoRefresh();
      }).not.toThrow();
    });

    it('should clear existing interval before starting new one', () => {
      const vm = wrapper.vm;
      const originalInterval = vm.autoRefreshInterval;

      vm.startAutoRefresh();

      expect(vm.autoRefreshInterval).not.toBe(originalInterval);
    });
  });

  describe('Error handling', () => {
    it('should handle missing dashboard store gracefully', () => {
      wrapper = createWrapper({ updateLastUpdatedRequest: null });

      expect(wrapper.exists()).toBe(true);
    });

    it('should handle component unmount during loading', async () => {
      const vm = wrapper.vm;

      const loadPromise = vm.loadData();
      wrapper.unmount();

      await loadPromise;
      expect(vm.isLoading).toBe(false);
    });
  });

  describe('Constants and configuration', () => {
    it('should have correct auto refresh interval', () => {
      const vm = wrapper.vm;
      expect(vm.AUTO_REFRESH_INTERVAL).toBe(60 * 1000);
    });
  });
});
