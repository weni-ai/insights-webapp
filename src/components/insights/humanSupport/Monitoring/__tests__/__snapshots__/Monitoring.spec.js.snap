// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Monitoring > Component structure > should match snapshot 1`] = `
<section
  class="monitoring"
  data-testid="monitoring"
  data-v-ada4682d=""
>
  <service-status-stub
    data-testid="monitoring-service-status"
    data-v-ada4682d=""
  />
  <time-metrics-stub
    data-testid="monitoring-time-metrics"
    data-v-ada4682d=""
  />
</section>
`;
